### PR TITLE
Record database metadata

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Record the settings that were used when the database was created to a `metadata.yml` file which is reloaded allowing us to change the values for new databases without having to introduce a new database version.

To make experimentation with different database configuration easier, values which can be changed without restarting the database can be added to the `metadata.yml` file (manually) and they will be loaded.  This avoids the need to add a bunch of `-X` options to the command line.

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

**Comment by ajsutton** [comment]
Will need to change this to not use ZSTD compression as it's proving unstable (or possibly just not reliably supported):
```
{"timestamp":"2020-06-26T07:11:37,217","level":"INFO","thread":"vert.x-eventloop-thread-1","class":"MetricsHttpService","message":"Metrics service started and listening on 8008:8008","throwable":""}
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fe9a9866fd2, pid=11463, tid=11547
#
# JRE version: OpenJDK Runtime Environment (11.0.2+9) (build 11.0.2+9)
# Java VM: OpenJDK 64-Bit Server VM (11.0.2+9, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C  [librocksdbjni6318734979621368944.so+0x5d9fd2]  ZSTD_decompressBegin+0x2
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /opt/teku/teku-0.11.5/hs_err_pid11463.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
```

**Comment by ajsutton** [comment]
I've removed the change to enable compression and just focus on the metadata change here.  Will run some experiments with different compression options to see what works (and with this change it no longer needs code changes!)

**Comment by ajsutton** [comment]
For the record the snyk failure is a false alarm.  We only include the fixed version of the dependency it's referring to and there are actually no new dependencies or changes to dependency versions in this PR (it just adds rocks db to a new module and snyk isn't very smart).

File: storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbConfiguration.java [hunk 1]
```diff
@@ -13,46 +13,86 @@
...
+  @JsonProperty(value = "writeBufferCapacity", access = Access.WRITE_ONLY)
+  private long writeBufferCapacity = DEFAULT_WRITE_BUFFER_CAPACITY;
+
+  /* ---------------     Fixed Properties     ------------ */
+
+  @JsonProperty("compressionType")
+  private CompressionType compressionType = CompressionType.NO_COMPRESSION;
+
+  @JsonProperty("bottomMostCompressionType")
+  private CompressionType bottomMostCompressionType = CompressionType.NO_COMPRESSION;
```
Comment (rolfyone) [inline storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbConfiguration.java:73]: not the hugest fan of bottomMost... lowest / minimum?

**Review by rolfyone — APPROVED** [review]
LGTM - one minor comment about 'bottomMost'

File: storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbConfiguration.java (hunk 1 above)
Comment (ajsutton) [reply storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/RocksDbConfiguration.java:73]: It's a RocksDB term - the DB is conceptually made up of a series of layers stacked on top of each other with data gradually being pushed down to the bottom layer as it's been unchanged for longer periods.  Hence, bottom-most.

**Pull request merged**